### PR TITLE
fs: Add the 'globs' option

### PR DESCRIPTION
**Pull request by cdlscpmv**

### CODE REVIEW

File: plugins/fs/fs.c [hunk 1]
```diff
@@ -108,10 +109,14 @@ run(LuastatusPluginData *pd, LuastatusPluginRunFuncs funcs)
         lua_State *L = funcs.call_begin(pd->userdata);
         lua_createtable(L, 0, p->paths.size);
         for (size_t i = 0; i < p->paths.size; ++i) {
-            const char *path = p->paths.data[i];
-            if (push_for(pd, L, path)) {
-                lua_setfield(L, -2, path);
+            glob_t gbuf = {0};
+            glob(p->paths.data[i], 0, NULL, &gbuf);
```
Comment (shdown) [inline plugins/fs/fs.c]: Why don’t you check the return value of `glob()`?

You would probably want to pass `GLOB_NOCHECK` and `GLOB_NOSORT` flags.

File: plugins/fs/fs.c [hunk 2]
```diff
@@ -108,10 +109,14 @@ run(LuastatusPluginData *pd, LuastatusPluginRunFuncs funcs)
         lua_State *L = funcs.call_begin(pd->userdata);
         lua_createtable(L, 0, p->paths.size);
         for (size_t i = 0; i < p->paths.size; ++i) {
-            const char *path = p->paths.data[i];
-            if (push_for(pd, L, path)) {
-                lua_setfield(L, -2, path);
+            glob_t gbuf = {0};
```
Comment (shdown) [inline plugins/fs/fs.c]: This ` = {0}` thing isn’t needed.

**Comment by shdown** [comment]
1. What’s the use case? If the list of matches does not change, it can be done on the Lua side.

2. So, if `paths` contain non-existing (or non-accessible) paths, it will silently swallow the errors? E.g. `opts = {paths = {'/', '/home', '/asdfdasfdasfsadfdssdf'}}`

File: plugins/fs/fs.c (hunk 1 above)
Comment (cdlscpmv) [reply plugins/fs/fs.c]: I don't check the return value because it may return only three kind of errors:
1. GLOB_NOSPACE - the system is out of memory. We can abort here, but I don't think this is really necessary. I can add the check if you want.
2. GLOB_ABORTED - the user has no permission to a directory, so it will be skipped. This is useful if the user have no access to certain directories (they will be silently ignored), but have access to the rest of them. I can add the check. Another option would be to specify the mentioned behavior in the docs.
3. GLOB_NOMATCH - because sometimes it is pretty legitimate for a glob not to match anything. See the comment below.

**Comment by cdlscpmv** [comment]
Let me describe the use case. I mount removable media in directories inside `/media`. The names
of these directories change from one device to another. So I wanted to display them in the statusline when they are plugged and hide them when they aren't. When no divice is plugged, the `/media` directory is empty which is why I ignored the GLOB_NOMATCH error.

If you can think of a better solution to my problem, I would like to hear.

File: plugins/fs/fs.c (hunk 1 above)
Comment (shdown) [reply plugins/fs/fs.c]: > GLOB_NOSPACE - the system is out of memory. We can abort here, but I don't think this is really necessary. I can add the check if you want.

Yeah, this definitely should be checked. Also, systems may invent their own reasons for `glob()` to fail (e.g. now-obsolete `GLOB_NOSYS`).

`GLOB_ABORTED` is impossible if `errfunc == NULL && (flags & GLOB_ERR) == 0`.

**Comment by cdlscpmv** [comment]
I've been thinking about how to differentiate between a non-existent path and a failing glob pattern. And I see no easy way to do that that except for writing almost the whole spec of glob expansion rules.

I'm thinking about introducing a separate options like `paths_glob` with the feature that patterns may expand to nothing. What are your thought on that?

**Comment by shdown** [comment]
> About GLOB_NOSORT, I thought that predictable order of displayed paths would be preferrable for a statusline.

The order in which `glob()` returns the results has no effect, as the results are then used as Lua table keys. The order of traversal of Lua’s `pairs()` is undefined.

> which is why I didn't pass the GLOB_NOCHECK flag.

Consider adding `globs` option in addition to `paths`.

**Comment by cdlscpmv** [comment]
> Consider adding globs option in addition to paths.

Will do!

**Comment by cdlscpmv** [comment]
I've implemented the new option. Regarding the code, I do not check for `GLOB_NOMATCH` because according to POSIX `gl_pathc` will be zero in this case (according to my interpretation).

File: plugins/fs/fs.c [hunk 3]
```diff
@@ -106,13 +122,26 @@ run(LuastatusPluginData *pd, LuastatusPluginRunFuncs funcs)
     while (1) {
         // make a call
         lua_State *L = funcs.call_begin(pd->userdata);
-        lua_createtable(L, 0, p->paths.size);
+        lua_newtable(L);
```
Comment (shdown) [inline plugins/fs/fs.c]: Why this change? There will be at least `p->paths.size` hash entries in the table.

File: plugins/fs/fs.c [hunk 4]
```diff
@@ -43,18 +49,28 @@ init(LuastatusPluginData *pd, lua_State *L)
...
+    PU_MAYBE_VISIT_TABLE_FIELD(-1, "globs", "'globs'",
+        PU_CHECK_TYPE(LS_LUA_KEY, "'globs' key", LUA_TNUMBER);
+        PU_VISIT_STR(LS_LUA_VALUE, "'globs' element", s,
+            LS_VECTOR_PUSH(p->globs, ls_xstrdup(s));
+        );
+    );
+
+    if (!p->paths.size && !p->globs.size) {
+        LS_WARNF(pd, "both paths and globs are empty");
+        goto error;
```
Comment (shdown) [inline plugins/fs/fs.c]: This is a warning, not an error.

File: plugins/fs/fs.c [hunk 5]
```diff
@@ -106,13 +122,26 @@ run(LuastatusPluginData *pd, LuastatusPluginRunFuncs funcs)
...
         for (size_t i = 0; i < p->paths.size; ++i) {
             const char *path = p->paths.data[i];
             if (push_for(pd, L, path)) {
                 lua_setfield(L, -2, path);
             }
         }
+        for (size_t i = 0; i < p->globs.size; ++i) {
+            glob_t gbuf;
+            int err = glob(p->globs.data[i], GLOB_NOSORT, NULL, &gbuf);
+            if (err == GLOB_NOSPACE) {
```
Comment (shdown) [inline plugins/fs/fs.c]: Rather,
```c
switch (glob(p->globs.data[i], GLOB_NOSORT, NULL, &gbuf)) {
case 0:
case GLOB_NOMATCH:
    break;
default:
    LS_WARNF(pd, "glob() failed (out of memory?)");
}
```

Remember systems may invent their own reasons for `glob()` to fail (although this is not too probable to do more sophiscated error reporting).

File: plugins/fs/fs.c [hunk 6]
```diff
@@ -106,13 +122,26 @@ run(LuastatusPluginData *pd, LuastatusPluginRunFuncs funcs)
...
+                ls_oom();
+            }
+            for (size_t n = 0; n < gbuf.gl_pathc; ++n) {
```
Comment (shdown) [inline plugins/fs/fs.c]: Please name this variable `j`, not `n`.

File: plugins/fs/fs.c (hunk 3 above)
Comment (cdlscpmv) [reply plugins/fs/fs.c]: Well, I just thought that since the number of glob matches may be anything, it is not that useful to preallocate a fixed number of entries. But, I think you are right -- it won't hurt if we do.

**Comment by cdlscpmv** [comment]
Thank you for guidance. At this point you wrote the whole code yourself :).

**Comment by shdown** [comment]
Thanks for the contribution!

> At this point you wrote the whole code by yourself =).

I remember how my first contribution to strace required 12+ rounds of review :)
